File: algo_basic_python_java/w9.-sorting/532.Reverse_Pairs/532_reverse_pairs.java
public class Solution {
    /**
     * @param A: an array
     * @return: total of reverse pairs
     */
    public long reversePairs(int[] A) {
        // write your code here
        if (A == null || A.length == 0) {
            return 0;
        }
        
        int[] tmp = new int[A.length];
        return mergeSort(A, 0, A.length - 1, tmp);

    }

    private int mergeSort(int[] A, int start, int end, int[] tmp) {
        
        if (start >= end) {
            return 0;
        }
        int cnt = 0;
        int mid = start + (end - start)/2;

        cnt += mergeSort(A, start, mid, tmp);
        cnt += mergeSort(A, mid + 1, end, tmp);
        cnt += merge(A, start, end, tmp);

        return cnt;
    }

    private int merge(int [] A, int start, int end, int[] tmp) {
        int n = end - start + 1;
        int mid = start + (end - start)/2;
        int left = start, right = mid + 1;
        int cnt = 0;

        for (int k = 0; k < n; k++) {
            if ((left <= mid) && (right > end || A[left] <= A[right])) {
                tmp[k] = A[left];
                left++;
            } else {
                tmp[k] = A[right];
                right++;
                if(left <= mid) {
                   cnt += (mid - left + 1); 
                }
            }
        }

        for (int k = 0; k < n; k++) {
            A[start + k] = tmp[k];
        }

        return cnt;
    }
}

File: lintcode_algo_basic_python_java/s9.-sorting/532.Reverse_Pairs/532_reverse_pairs.py
class Solution:
    """
    @param A: an array
    @return: total of reverse pairs
    """
    def reversePairs(self, A):
        # write your code here
        if not A:
            return 0
        tmp = [0 for _ in range(len(A))]
        return self.merge_sort(A, 0, len(A) - 1, tmp)
    
    def merge_sort(self, A, start, end, tmp):
        if start >= end: 
            return 0
        mid = start + (end - start)//2
        # record reversed paris num
        cnt = 0
        # left and right parts merge
        cnt += self.merge_sort(A, start, mid, tmp)
        cnt += self.merge_sort(A, mid + 1, end, tmp)

        # merge two sorted arrays
        cnt += self.merge(A, start, end, tmp)
        return cnt
    
    def merge(self, A, start, end, tmp):
        mid = start + (end - start)//2
        left, right = start, mid + 1
        n = end - start + 1
        
        #record reversed paris num
        cnt = 0
        for k in range(n):
            if left <= mid and (right > end or A[left] <= A[right]):
                tmp[k] = A[left]
                left+=1
            else:
                tmp[k] = A[right]
                right+=1
                if left <= mid:
                    cnt += mid - left + 1
        
        for k in range(n):
            A[start + k] = tmp[k]
        return cnt